docroot/sites/uipress.uiowa.edu/modules/uipress_core/uipress_core.module

```diff
@@ -141,25 +142,31 @@ function uipress_core_preprocess_paragraph__book_type(&$variables) {
         // which can be used to compare the start and end dates.
         $timezone = date_default_timezone_get();
         $datetime = new DrupalDateTime('now', $timezone);
-        $today = $datetime->format('m-d-Y 00:00:00');
+        $today = $datetime->format('Y-m-d 00:00:00');
 
         // If the start date exists make sure that it is in the past or today.
         $start = $paragraph->get('field_book_sale_start_date')->value;
         if (isset($start)) {
-          $start = date('m-d-Y', strtotime($start));
+          $start = date('Y-m-d', strtotime($start));
           $in_range = $today >= $start;
         }
 
         // If end date exists make sure it is after today.
         $end = $paragraph->get('field_book_sale_expiry_date')->value;
         if ($in_range && isset($end)) {
-          $end = date('m-d-Y', strtotime($end));
+          $end = date('Y-m-d', strtotime($end));
           $in_range = $today < $end;
         }
         // If neither condition is true unset display of sale price.
         if (!$in_range) {
           unset($variables['content']['field_book_sale_price'], $variables['content']['field_book_sale_code']);
         }
+        // If we are in range of a sale, invalidate the paragraph cache
+        // to make sure we're updating the display.
+        if ($in_range) {
+          $tags = $paragraph->getCacheTagsToInvalidate();
+          Cache::invalidateTags($tags);
+        }
```

Comment:
Would love to go over this or other options @joewhitsitt. I think this'd get the job done (make sure the paragraph is being rendered fresh if a sale has started), but feels off to me.

Reply:
I'm a little uncertain about this approach. I'm trying to reason about when it will be triggered and it seems like it would be triggered when the cache for the paragraph is already invalidated, since the preprocess function should only run when the paragraph needs to be re-rendered because the cache is invalid. I may be wrong about this assumption.

Reply:
Eh...I think you're right about that. I'd already felt off about it, so I think it definitely needs to be at least moved elsewhere, if not handled differently.

Reply:
What I was/am hoping to avoid is invalidating the cache for all books that may or may not have sales, even if it's somewhat mitigated by being handled through a cron job. But the more I think about it, the more I'm unsure about how to guarantee that the sale logic will trigger this invalidation. Maybe I'm just overthinking it and getting confused.

Reply:
> Cron is going to be your trigger. That is pretty much all we have right now, however, it can be a uipress specific hook_cron (e.g. every morning) that does a entityQuery to only get nodes that meet the condition and clear those.

Okay! That makes a lot of sense, and sounds like what I was trying to get to but couldn't quite grok.

Reply:
Have a look at https://github.com/uiowa/uiowa/blob/main/docroot/modules/custom/uiowa_core/uiowa_core.module#L38. It technically runs every time, but you can set a different time interval and instead of the hourly cache tag you can queue up a list of nodes to clear instead which should cascade down and clear blocks.